stdlib/public/core/StringUTF8.swift

```diff
@@ -240,20 +259,30 @@ extension String {
       internal static var _bufferHiByte: Buffer {
         return 0xFF << numericCast((sizeof(Buffer.self) &- 1) &* 8)
       }
+      
+      /// Amounts to increment the UTF-16 index based on the high nibble of a
+      /// UTF-8 code unit. If the high nibble is:
+      ///
+      /// - 0b0000-0b0111: increment 1
+      /// - 0b1000-0b1011: continuation character, increment 0
+      /// - 0b1100-0b1110: increment 1
+      /// - 0b1111: UTF-16 surrogate pair, increment 2
```

Comment:
I don't think this is quite right—the byte in question is from the UTF-8 encoding of the string, not the UTF-16 encoding, so we will see lots of continuation bytes. Any scalar with a code point > 127 will be have between one and three continuation bytes in its UTF-8 encoding. How about:

```
- 0b0000-0b0111: U+0000...U+007F: increment the UTF-16 pointer by 1
- 0b1000-0b1011: UTF-8 continuation byte, do not increment the UTF-16 pointer
- 0b1100-0b1110: U+0080...U+FFFF: increment the UTF-16 pointer by 1
- 0b1111:        U+10000...U+1FFFFF: increment the UTF-16 pointer by 2
```